### DISCUSSION TITLE
Alternative activation using mouse button?

### DISCUSSION THREAD

**Discussion by peterept** [rxhanson/Charmstone-Community#8] [Ideas]
Hi, Would love to map this to Mouse Button 3, or other mouse button?

**Comment by rxhanson** [comment]
You can configure it with this Terminal command:

`defaults write com.knollsoft.Charmstone reticleMouseButton -int 3`

Restart the app after executing it. 

Next time I rework the UI, I'll consider adding it there.

**Reply by peterept** [reply]
Wow, this is so amazing. Thanks Ryan. I've been using it all day, and middle button is so useful for Charmstone!

**Reply by peterept** [reply]
3 week update: 

So I love middle mouse button a lot, except for when it conflicts with other programs:
- Middle mouse button is used in several graphical programs I use 
- Middle mouse button performs paste in terminal

Can I ask for an extra feature:  Ability to silently cancel the operation if I hit any key other key (eg: shift, or esc). That way if I'm using the graphical program and I press the middle mouse button, I want to cancel charmstone - so pressing a key could cancel it and I keep performing the operation in 3d program.

I don't see a solution for pasting in terminal, I want to find a way to disable middle button paste. I'll find the separately from charmstone.

Thanks so much, this is so useful!

**Reply by rxhanson** [reply]
Thanks for the update. 

> Ability to silently cancel the operation if I hit any key other key (eg: shift, or esc)

I'll see what I can do, but it it might take a little while before I can get to it.

**Reply by arham-saeed** [reply]
Hey @rxhanson . this command doesn't work for me for some reason. 
These are my properties.

`{
    NSNavLastRootDirectory = "~/Documents";
    NSNavPanelExpandedSizeForOpenMode = "{800, 448}";
    "NSStatusItem Visible Item-0" = 0;
    "NSWindow Frame NSNavPanelAutosaveName" = "356 472 800 448 0 0 1512 944 ";
    arrowKeysDisplays = 2;
    autoRestartOnWake = 1;
    farReticleSpec = "{\\"b\\":{\\"stringValue\\":\\"com.apple.iphonesimulator\\",\\"type\\":0},\\"id\\":1,\\"l\\":{\\"stringValue\\":\\"notion.id\\",\\"type\\":0},\\"t\\":{\\"type\\":0,\\"stringValue\\":\\"com.googlecode.iterm2\\"}}";
    firstRingDist = "171.2944";
    hideMenuBarIcon = 1;
    hoverMillis = 0;
    lastVersion = 31;
    launchOnLogin = 1;
    mainReticleSpec = "{\\"r\\":{\\"type\\":0,\\"stringValue\\":\\"com.google.Chrome\\"},\\"l\\":{\\"type\\":0,\\"stringValue\\":\\"company.thebrowser.Browser\\"},\\"bl\\":{\\"type\\":0,\\"stringValue\\":\\"com.apple.dt.Xcode\\"},\\"br\\":{\\"stringValue\\":\\"com.postmanlabs.mac\\",\\"type\\":0},\\"b\\":{\\"stringValue\\":\\"com.jetbrains.intellij\\",\\"type\\":0},\\"t\\":{\\"type\\":0,\\"stringValue\\":\\"com.tinyspeck.slackmacgap\\"},\\"tr\\":{\\"type\\":0,\\"stringValue\\":\\"us.zoom.xos\\"},\\"id\\":0,\\"tl\\":{\\"stringValue\\":\\"com.microsoft.VSCode\\",\\"type\\":0}}";
    outerSafeArea = "438.2736";
    relaunchActive = 1;
    restToOneTouches = 3;
    reticleDisplayMinDist = "2.334906";
    reticleFlags = 524288;
    reticleMouseButton = 10;
    safeAreaSize = 11;
}
`

**Reply by rxhanson** [reply]
@arham-saeed are you able to get this working with a different mouse button, say, 3 for the middle button? Just trying to narrow it down and see if there's a mismatch between what you have configured as button 10 vs what Charmstone thinks is button 10.

**Reply by arham-saeed** [reply]
@rxhanson I'm using logitech mouse with some custom actions mapped to buttons using their Logi+ app. I think that is casuing the issue since it's probably intercepting all the inputs before the charmstone app.

**Comment by chrisgrieser** [comment]
somewhat related, I was wondering whether there some some method for triggering charmestone solely with the trackpad? would probably be a bit more tricky since there isn't really any mouse button on that 🤔

**Reply by rxhanson** [reply]
I've gone back and forth over this quite a bit. I have had plans to add what's in Rectangle Pro, the rest N, remove all but one gestures, but there are some downsides to doing it and it's more of a weighty decision than appears on the surface.

Right now, Charmstone needs no extra privileges and is sandboxed. In order to listen to touch data, I would at the very least need the user to give it Input Monitoring capabilities (Accessibility capabilities cover this in Rectangle Pro). If I were to go down this path, it would likely be more compelling to just create a non-App Store version that has more "Pro" features like this where I have the freedom to do so.

**Reply by chrisgrieser** [reply]
My thinking is just that charmstone is an amazing app switcher/launcher, when you have many apps open even quicker than alt+tab. When I have my hands on the keyboard, I already have tons of options for switching/launching apps, but when I have my hand at the mouse, I barely have any options other than Launchpad or some complicated BetterTouchTool scripting. So as I saw it, having a mouse-only app switcher was the game changer regarding charmestone. With your suggested Terminal command, charmstone can be used with mouse only, but for the trackpad, I don't see a method yet to use solely that one device.

I mean, I see where you are coming from, and I mean, in the end it's your vision of the product of course, but maybe my train of thought shows how trackpad/mouse only would make charmstone fill a gap, in my view?

**Reply by rxhanson** [reply]
I appreciate the continued conversation here. Another element in this decision is that there is a small subset of users that tend to always have issues with touch gestures - I've had some users of Multitouch that have experienced the touch gesture API from macOS just not working at all, or stopping working at some point randomly, or being affected by other apps. Charmstone by contrast, has almost no bugs reported (albeit with a much, much smaller user base). Even if there's a small number of users that have issues with the touch API, it is both difficult to figure out the root cause of such issues, and it's a bit disheartening from a developer perspective to feel like something just won't work right for everyone, regardless of anything that I can try from my side.

In regard to trackpad gestures, I really like the rest N, remove all but one gestures in Rectangle Pro for this, but I really haven't received much feedback for people actually using it. Is this something you would use/prefer in Charmstone, or do you have other thoughts on a trackpad trigger? I like the rest N, remove all but one so much that I would consider building an app around just that gesture, but I don't know if anyone else would want that but me :)

I'll give this some more thought though. I've come really close to prioritizing this up and implementing it before. Right now I'm working on a "pro" version of Hyperkey and will see if that's what I want to do next after that's out the door.

**Reply by chrisgrieser** [reply]
> In regard to trackpad gestures, I really like the rest N, remove all but one gestures in Rectangle Pro for this, but I really haven't received much feedback for people actually using it. Is this something you would use/prefer in Charmstone, or do you have other thoughts on a trackpad trigger?

I just installed Rectangle Pro to try it out, but I can't seem to find it 😅 
Could you point me to where the feature is?

**Reply by rxhanson** [reply]
![image](https://user-images.githubusercontent.com/13651296/170827251-2f2414a6-2e67-44e0-b490-6ebd90f6c423.png)

I prefer "Rest 4 Touches, Remove All Except 1"

**Reply by chrisgrieser** [reply]
hmmm, the gesture takes a bit practice to get used to compared to things like pinching or double-tapping with 3 fingers etc. But after getting used to it, it does feel quite smooth. 

The idea is to have a gesture where you still end up with a finger on the trackpad for moving, is it? That's quite nifty, and I think it would be a wonderful method for triggering charmestone.

**Reply by rxhanson** [reply]
Thanks for trying it out and letting me know! I’m still not sure which path I want to take (pro/non-mas version or not), but I’ll plan on adding it at some point. There’s still the chance that it might not be sandboxable and forced to be outside of the MAS anyway. If it ends up being a separate app, I would likely give it away free to current full version owners.

**Reply by rxhanson** [reply]
Many months later, I finally did end up adding the trackpad gestures in from Rectangle Pro, and I use it all the time. Wish I woulda done it sooner.

**Comment by hugows** [comment]
Still wishing I could activate Charmstone with the mouse only.. any ideas? I use Steermouse but that didn't help. Could there a mode where Charmstone appeared when the app was "opened" instead of when the mouse moves with the modifiers pressed?
Then I could map any button chord to Charmstone...

**Reply by rxhanson** [reply]
You can configure a mouse button to it using this terminal command (shown at the top of this thread):

`defaults write com.knollsoft.Charmstone reticleMouseButton -int 3`

Since the app is sandboxed, it can't prevent mouse button presses from propagating through the rest of macOS, and this is why I don't expose it in the UI. For example, mouse buttons 4 & 5 will perform back & forward in Chrome and there's no way to disable that - leading to confusing behavior if those buttons are chosen.

> Could there a mode where Charmstone appeared when the app was "opened"

I'll think about this. Perhaps it would be more desirable just to have urls for scripting close and open? for example `charmstone://open` and `charmstone://close`?

**Reply by hugows** [reply]
Yes, something like that! Would that still allow to open the action at the cursor position?

**Reply by rxhanson** [reply]
Yes. I'll add it to my queue. Did you try out the terminal command? That should get what you want in the meantime.

**Reply by hugows** [reply]
Yes, that worked for me - maybe I'll just try that. I'll have to disable middle click to paste on iTerm though.
Edit: thats so fast...

**Reply by hugows** [reply]
Unfortunately it also conflicted/broke VSCode too...

https://user-images.githubusercontent.com/11545/214107495-5f472c27-a6e4-4eef-87a6-9ab376afc7c3.mov

**Reply by rxhanson** [reply]
Ah, good to know. Well, I'll post back here when the url scripting is ready.

**Reply by NeighNeighNeigh** [reply]
> Perhaps it would be more desirable just to have urls for scripting close and open? for example `charmstone://open` and `charmstone://close`?

I came here looking for this, haha, so +1

**Reply by rxhanson** [reply]
This just got rolled into the last release. I got sidetracked and didn't post it. Here are the possible urls at the moment:

`charmstone://show`

`charmstone://close`

`charmstone://close?executing=false`

Let me know how it goes.

**Reply by NeighNeighNeigh** [reply]
> Let me know how it goes.

Hmm, doesn't seem to work as expected. Opening `charmstone://show` does not open the main Charmstone UI. If the Charmstone setting panel happens to be open, running the URI will focus the Charmstone setting panel. `charmstone://close` does nothing.

**Reply by rxhanson** [reply]
Try restarting the app, or restarting your Mac. Sometimes macOS doesn't register the urls.

**Reply by NeighNeighNeigh** [reply]
I found the problem, it won't work if `Min Dist to Appear` is non zero (which is my preferred configuration). Would be great if the URI launch would occur regardless of this.

**Reply by rxhanson** [reply]
Ah, good to know, thanks. I'll see what I can do for the next release.

**Reply by patrickrushton** [reply]
charmstone://show works after setting Min Dist to Appear to zero, but the Charmstone UI disappears as soon as I move the cursor. Is this the expected behaviour?

**Reply by rxhanson** [reply]
@patrickrushton, thanks for bringing this back up, hopefully I can get to this soon. It's not intended behavior.

**Reply by patrickrushton** [reply]
That would be awesome. I was hoping to be able to invoke Charmstone from my Wacom tablet, so it would be really helpful to get this working.

**Reply by rxhanson** [reply]
In the latest version of Charmstone, v0.30, I added a query parameter to the show url, and this allows you to specify an "appearDist" that will override the Min Dist to Appear setting.

I think this should cover the scenarios here, let me know if something's not right.

```
charmstone://show?appearDist=0
```

**Reply by NeighNeighNeigh** [reply]
Hmm, I just installed installed the latest version and launched Charmstone using `charmstone://show?appearDist=0` which brought up the Charmstone window, but as soon as I moved the mouse, the window disappeared. 
Here are my current Charmstone settings:
![SCR-20230822-klpy](https://github.com/rxhanson/Charmstone-Community/assets/4795315/d7b0c56a-f1e1-402b-bf81-3b843d670123)
![SCR-20230822-klqp](https://github.com/rxhanson/Charmstone-Community/assets/4795315/6e41ca96-7a2b-48a3-bd4f-c9b8ec5c7199)

**Reply by rxhanson** [reply]
@NeighNeighNeigh thanks for trying it out quickly and reporting back. What are you using to open the URL?

**Reply by NeighNeighNeigh** [reply]
I'm trying to make it work with Alfred like so:
![SCR-20230822-kxoa](https://github.com/rxhanson/Charmstone-Community/assets/4795315/ee9d528f-92b8-4bf1-b3db-a331e76e665b)

but I've also tried simply clicking on the url in a textedit RTF document:
https://github.com/rxhanson/Charmstone-Community/assets/4795315/93f64cdc-2ab5-448d-ac60-15305cd1005d

**Comment by katzelad** [comment]
Thanks for the app.
Is there any way to activate using only the magic mouse? For example by right clicking.

**Reply by rxhanson** [reply]
Not at the moment, at least not on its own. You could use one of my other apps, [Multitouch](https://multitouch.app) or [Middle](https://middleclick.app) to add middle clicking to the Magic Mouse, then configure Charmstone with the following terminal command: `defaults write com.knollsoft.Charmstone reticleMouseButton -int 3`, and that would do it. I'll make note of this for a future release - nobody's asked for it before.

**Comment by leocabrallce** [comment]
Hey! Same thing here... This project is awesome, but I do have a mouse with a few extra buttons (Logitech G502) and I would really love to use one of those extras as the trigger to Charmstone. The problem is that I don't know how to properly config this extra button as the trigger. Just for context, the DPI Shift side button.
I tried configuring the corresponding button as a Macro, simulating the press of the configured keys (CMD + Alt) on Charmstone, but this doesn't work either, even if I configure Logitech's system to simulate holding the buttons.

**Reply by rxhanson** [reply]
Logitech's software likely doesn't propagate the modifiers to the mouse moved and drag event types.

If you can figure out or assign the mouse button number of this button, then you can use this terminal command:

```
defaults write com.knollsoft.Charmstone reticleMouseButton -int <BUTTON_NUMBER>
```

If you're using Charmstone outside of the Mac app store, then the bundleID is different:

```
defaults write com.knollsoft.CharmstonePro reticleMouseButton -int <BUTTON_NUMBER>
```

Restart the app after executing this.

**Comment by wiesnero** [comment]
Hey,

I just bought the app as soon as I discovered that the URLs mentioned here exist, allowing me to create a Mouse Shortcut. I was considering writing a support email before I found this thread. It would be helpful to include this information directly in the app, as I was close to abandoning it. 

After creating a Shortcut for the "charmstone://show?appearDist=0" link with Raycast, I encountered the same issue mentioned earlier: Charmstone immediately disappears when I move my mouse.

However, this only happens when Trackpad Gestures are turned off. If you enable them, Charmstone remains active, and you can select an app with the mouse as expected. So, for anyone experiencing the same issue, try turning on Trackpad Gestures. It seems there is a bug related to this.

**Reply by rxhanson** [reply]
Good to know, thanks! I'll see what I can do.

**Reply by enieuwy** [reply]
@wiesnero how did you assign the mouse button as a hotkey in Raycast? I've made a quicklink—it works, but I can't seem to assign the middle mouse button as a hotkey (Logitech MX Anywhere 3S).

**Reply by wiesnero** [reply]
@enieuwy in Raycast i have only configured the quicklink. 
I use bettermouse to assign my my mouse button to the quicklink.

**Reply by olifrost** [reply]
This is also my setup, except I'm using Keyboard Maestro for the quick link

**Reply by enieuwy** [reply]
Awesome, thanks @wiesnero. I've installed BetterMouse (very happy with it—for me, it also replaced Mos and Logi Options), and was able to get this working. Your note about the trackpad gestures also saved much head scratching.

**Reply by olifrost** [reply]
@enieuwy What are you putting in the quick link? It hasn't been working for me

**Reply by enieuwy** [reply]
I'm using `opt + cmd + enter` for the hotkey and `charmstone://show?appearDist=0` for the link

**Reply by olifrost** [reply]
@enieuwy Ah but I find that it dismisses as soon as I move the mouse

**Reply by enieuwy** [reply]
That's where wiesnero's original tip comes into play—in Charmstone, you need to set a trackpad gesture (until this is fixed I presume).

**Reply by rxhanson** [reply]
This should now be fixed in the latest version (1.37). The App Store version has been submitted to the App Store.

**Comment by dgmulf** [comment]
Middle click feels very natural when using a traditional mouse. I think a toggle in the GUI settings would be great.

**Comment by olifrost** [comment]
Having trouble with all the approaches here with a Logitech mouse.
Charmstone dismisses itself when mouse is moved after the URL scheme 
Tried changing the hidden preference and restarting, but none of the Logitech buttons seem to be recognised
A keyboard shortcut or Applescript trigger would be simple, so one could chain it together with Keyboard Maestro or another app

**Reply by enieuwy** [reply]
Also can't get this to work.

**Reply by rxhanson** [reply]
I'd like to finally get URLs working for every setup. I'll post back when I have something to test out. I don't mind adding in keyboard shortcut equivalents for show/hide, but it would likely have the same issue as the URL so I'll plan on getting that working first.

**Reply by rxhanson** [reply]
This should now be fixed in the latest version (1.37). The App Store version has been submitted to the App Store.

**Reply by noahkiss** [reply]
I didn't have any luck with Logi Options+ and spent a long time trying all different combinations (including launching the URL). I had to settle on a combo of SteerMouse and Mos just to pass through the mouse button I wanted (MX Master 3, button 7 which is the scroll toggle button). 

I'd prefer to go back to Logi O+, though, and a keyboard shortcut would most likely fix these issues - Logi O+ has no problem with keyboard shortcuts, which are first-class as opposed to app url schemes. Unfortunately, Logi O+ didn't work with just modifier keys as the shortcut - it allows setting a button to a shortcut of just ctrl+opt+cmd+shift or any combo, but doesn't send the keypresses in the right way for Charmstone to notice them, I guess. 

Definitely looking forward to a keyboard shortcut combo, I think it's the lowest friction / most tried-and-true way of allowing for more options for launching. URL schemes and just modifier keys are nice, but are harder to implement in a lot of places. I wouldn't be surprised if there are accessibility cases where a normal kb shortcut would be a better choice, too.

**Reply by rxhanson** [reply]
@noahkiss thanks for the details, I'll give the keyboard shortcut a bump in priority. I'm assuming that the desired functionality would be that a keyboard shortcut down/up would open Charmstone and another would close/select it? Rather than keyboard shortcut down to open Charmstone and up to close/select it.

**Reply by noahkiss** [reply]
I'm probably not the best person to answer that, I'm still fairly new to Charmstone and haven't really developed preferences yet. Right now, I have it set up via Raycast as a quicklink to the`charmstone://show?appearDist=0` mapped to a kb shortcut, and mapping that to my mouse button requires me to hold down the button, drag to app, and let go of button. I think I like it that way as it's less presses than click to open, aim, and click (in my case, a different button) to select. It feels like one of those things where different people would have different preferences, though, so I don't want to speak out of turn.

**Comment by enieuwy** [comment]
Latest release seems to have broken my middle-mouse function. I used to hold the button and release on the selection, but now Charmstone doesn't select the app on release (I have to click).

I've bound the button to: charmstone://show?appearDist=0

**Reply by rxhanson** [reply]
Thanks for letting me know, I think I see what caused this and will have a fix out soon.

**Reply by enieuwy** [reply]
Appreciate the quick action @rxhanson!
I'm not sure if the fix introduced a new bug though—sometimes I have a "ghost" menu remain after an app is opened. The selector shows but is unresponsive until I press middle-mouse a second time.

**Reply by rxhanson** [reply]
Thanks for following up. What are you using to map a mouse button to the url? The fix in v1.39 effectively rolls back the change in 1.38, and that change in 1.38 was an adjustment for modifier flag changes (which I would have expected to not cause issues with URL triggering).

**Reply by enieuwy** [reply]
I've assigned the hotkey with BetterMouse.
I think I've found the issue only occurs when I'm using [AutoRaise](https://github.com/sbmpost/AutoRaise)

**Reply by rxhanson** [reply]
In your url execution, it sounds like you only have the url configured for opening Charmstone? If so, then the behavior that you had previously was likely only working by chance of the events being pushed into the event stream by other apps. I'm always cautious of trying to work around other apps that modify behavior like this, as it can be significant effort to line everything up just right and keep it lined up. In this case, the behavior on Charmstone's side is effectively unchanged.

Is it possible to configure mouse down to the open url and mouse up to the close url? Would it be better to configure Charmstone to respond to keyboard shortcuts, so that the mouse down and mouse up correspond to showing and hiding Charmstone on down/up?

**Reply by enieuwy** [reply]
I was able to correct it by simply asking AutoRaise to ignore Charmstone.
Love using Charmstone with the middle mouse button: Button-down, move to desired app, button-up.

**Comment by enieuwy** [comment]
The latest release 1.43 (43) appears to have broken this function for me (in a new way). Using the modifier (opt+cmd) works normally, but using middle mouse (charmstone://show?appearDist=0 via Raycast) doesn't allow me to select and of the menu items (no hover-over effects either).

**Reply by rxhanson** [reply]
Thanks for reporting. Nothing's changed in that part the code for a number of releases. Do you know what the most recent release was where this worked for you? That way I can narrow it down.

**Comment by enieuwy** [comment]
Appreciate the quick response. I was on the latest version (don’t recall the number), so I’m fairly confident it occurred with the latest update (I update regularly and use it daily).

**Reply by rxhanson** [reply]
Perhaps I can reproduce it. How do you have charmstone://show?appearDist=0 configured in Raycast?

**Comment by rcongdv** [comment]
I seem to be having a similar issue, and don't see a solution on here that works for me. I can't get the extra buttons on my Logitech Anywhere 3S mouse to work. I do NOT use logi options+ and I am trying this with no other mouse extensions running. I installed with homebrew, the middle button (3) seems to work but I tried up to 10 and none of them seem to be working with the other buttons. The wheel just doesn't show up at all. Any way to see what numbers the buttons are mapped too? Couldn't find anything online.

works (middle button):
`defaults write com.knollsoft.CharmstonePro reticleMouseButton -int 3`

doesn't work:
`defaults write com.knollsoft.CharmstonePro reticleMouseButton -int 4`

**Reply by rxhanson** [reply]
From looking at images of the Logitech Anywhere 3S, my expectation is that those side buttons would be recognized by Charmstone as 4 and 5.

As a quick sanity check, did you restart the app after executing the Terminal command for 4?

I still have a backlog item for adding a button recorder to the UI, and I realize that this suggestion is annoying, but I do have a mouse button recorder in the Window Throw section of Rectangle Pro that would be able to tell you which button number it recognizes, which you would then configure in Charmstone:
<img width="821" height="620" alt="image" src="https://github.com/user-attachments/assets/0f8c5c46-cd5c-42ed-9451-57f152348c4c" />

The typical scenario for a button like 4 or 5 not being recognized is that another mouse utility is capturing the button click before it reaches Charmstone, but since you've mentioned that you don't have one running I'm not sure off the top of my head what could cause this.

**Reply by rcongdv** [reply]
Would you mind sharing the code you are using to capture this if its just a few lines? I tried this but it does not detect any of the extra buttons, only the middle click as 3, which is probably the issue

**Reply by rxhanson** [reply]
Charmstone & Rectangle Pro are reading events of type "otherMouseDown" to obtain this. Left and right mouse buttons are "leftMouseDown" and "rightMousedown", but all other mouse buttons are "otherMouseDown". In other words, the logic for obtaining the middle mouse button and the other buttons are the same, so your mouse appears to be designed to not send standard macOS events for the side buttons (without Logitech software running), despite being designed to send the standard middle button events. With that in mind, I don't have a straightforward solution to this, unless the mouse is sending some other kind of event that can be tracked.

**Reply by rcongdv** [reply]
Ok thanks for your reply. I will try the same mouse with some other machines, maybe I have some hidden mouse software running that I forgot about. Thanks!

**Comment by hdr-44** [comment]
Hi guys, I just managed to do this using Karabiner:

Triggering Charmstone with an MX Master 4 Middle-Click (Hold):
   1. Quick Tap: Acts as a normal middle-click (e.g., opening a link in a new tab).
   2. Hold + move: Triggers Charmstone

## Step 1: Enable Your Mouse in Karabiner
By default, Karabiner ignores mice to prevent cursor issues.

   1. Open Karabiner-Elements Settings.
   2. Go to the Devices tab.
   3. Find your MX Master (or "Logitech Device / USB Receiver") and check the box in the Modify events column. 

## Step 2: Configure Charmstone Shortcut

   1. Open Charmstone Settings.
   2. Set your activation shortcut. I use Command + Fn
   
## Step 3: Add the Custom Rule to Karabiner
In Karabiner go to Complex Modifications tab – "Add your own rule" – paste the code below:

```json
{
  "description": "Mouse Button 3: Tap for Middle-Click, Hold for Charmstone (Cmd+Fn)",
  "manipulators": [
    {
      "type": "basic",
      "from": {
        "pointing_button": "button3",
        "modifiers": { "optional": ["any"] }
      },
      "to_if_alone": [
        {
          "pointing_button": "button3"
        }
      ],
      "to_if_held_down": [
        {
          "key_code": "left_command",
          "modifiers": ["fn"],
          "repeat": false
        }
      ],
      "parameters": {
        "basic.to_if_held_down_threshold_milliseconds": 200
      }
    }
  ]
}
```
___
Feel free to modify milliseconds, Charm shortcut or mouse button.
